rendering/mmd_pmx.ksy

```diff
@@ -0,0 +1,569 @@
+  - id: frame_count
+    type: u4
+  - id: frames
+    type: frame
+    repeat: expr
+    repeat-expr: frame_count
+  - id: rigid_body_count
+    type: u4
+  - id: rigid_bodies
+    type: rigid_body
+    repeat: expr
+    repeat-expr: rigid_body_count
+  - id: constraint_count
+    type: u4
+  - id: constraints
+    type: constraint
+    repeat: expr
+    repeat-expr: constraint_count
+
+types:
+
+  header:
+    seq:
+      - id: magic
+        contents: 'PMX '
+      - id: version
+        type: f4
+      - id: header_size
+        type: u1
+      - id: encoding
+        type: u1
+      - id: additional_uv_count
+        type: u1
+      - id: vertex_index_size
+        type: u1
+      - id: texture_index_size
+        type: u1
+      - id: material_index_size
+        type: u1
+      - id: bone_index_size
+        type: u1
+      - id: morph_index_size
+        type: u1
+      - id: rigid_body_index_size
+        type: u1
+      - id: model_name
+        type: len_string
+      - id: english_model_name
+        type: len_string
+      - id: comment
+        type: len_string
+      - id: english_comment
+        type: len_string
+
+  vertex:
+    seq:
+      - id: position
+        type: f4_3
+      - id: normal
+        type: f4_3
+      - id: uv
+        type: f4_2
+      - id: additional_uvs
+        type: f4_4
+        repeat: expr
+        repeat-expr: _root.header.additional_uv_count
+      - id: type
+        type: u1
+      - id: skin_indices
+        type:
+          switch-on: _root.header.bone_index_size
+          cases:
+            1: u1
+            2: u2
+            4: u4
+        repeat: expr
+        repeat-expr: 'type == 0 ? 1 : type == 1 ? 2 : type == 2 ? 4 : type == 3 ? 2 : -1'
+      - id: skin_weights
+        type: f4
+        if: type == 1 or type == 2 or type == 3
+        repeat: expr
+        repeat-expr: 'type == 1 ? 1 : type == 2 ? 4 : type == 3 ? 1 : -1'
+      - id: skin_c
+        type: f4_3
+        if: type == 3
+      - id: skin_r0
+        type: f4_3
+        if: type == 3
+      - id: skin_r1
+        type: f4_3
+        if: type == 3
+      - id: edge_ratio
+        type: f4
+
+  face:
+    seq:
+      - id: indices
+        type:
+          switch-on: _root.header.vertex_index_size
+          cases:
+            1: u1
+            2: u2
+            4: u4
+        repeat: expr
+        repeat-expr: 3
+
+  texture:
+    seq:
+      - id: name
+        type: len_string
+
+  material:
+    seq:
+      - id: name
+        type: len_string
+      - id: english_name
+        type: len_string
+      - id: diffuse
+        type: f4_4
+      - id: specular
+        type: f4_3
+      - id: shininess
+        type: f4
+      - id: ambient
+        type: f4_3
+      - id: flag
+        type: u1
+      - id: edge_color
+        type: f4_4
+      - id: edge_size
+        type: f4
+      - id: texture_index
+        type:
+          switch-on: _root.header.texture_index_size
+          cases:
+            1: u1
```

Comment:
Again should be enums.

Reply:
The meaning of the value is "the number of bytes in the index". Usually when you have a number of things you want to use an int, but here you're suggesting an enum.

Perhaps it might be better to keep as int but declare `valid` for the index size stuff as only valid if 1,2,4?

Reply:
So I will add new types for each index type like this:

```yaml
  bone_index:
    doc: Variable-length type storing a bone index
    seq:
      - id: value
        type: switch-on: _root.header.bone_index_size
        cases:
          1: u1
          2: u2
          4: u4
```

Reply:
Now I have removed repetition in the structures above, and have instead 6 repetitive types which only differ in three places:
- their name
- the doc comment
- the field they use to switch on

Reply:
You can pass the condition in `switch-on` as a param, BTW.

Reply:
I think I tried that in this situation once in the past and it didn't work, but it will be tried.

Reply:
OK, I got it down to:

```yaml
      - id: indices
        type: sized_index(_root.header.vertex_index_size)
```

I think I'll call this _good enough_ but it would be nice if I could write

```yaml
      - id: indices
        type: sized_index(index_id::vertex)
```

But (a) I'd be declaring an enum with fictional values just to use it as a param and (b) I'm 99% sure that the last time I tried using enums in params they didn't work anyway.

Reply:
> But (a) I'd be declaring an enum with fictional values just to use it as a param and (b) I'm 99% sure that the last time I tried using enums in params they didn't work anyway.

They were added in the 0.9 release, see the [release notes](https://kaitai.io/news/2020/10/16/kaitai-struct-v0.9-released.html):

> - Major bugfixes:
>   - `params/type` - add support for:
>     - specific user types
>     - `enum` types (#413)
>     - byte arrays (`bytes`)
>     - arrays (`u2[]`, `struct[]`, etc.)